napari/plugins/_hook_callers.py

```diff
@@ -0,0 +1,272 @@
+"""Modified _hookexec caller.
+
+When using a pluggy PluginManager, when you actually call a hook using their
+syntax ``plugin_manager.hook.hook_spec_name(**kwargs)`` it goes through a
+relatively complicated chain of methods and aliases before ultimately landing
+on the basic function that loops through plugins: ``pluggy.callers._multicall``
+
+This module is here as a patch because:
+
+1. pluggy doesn't indicate WHICH hook implementation returned the result(s)...
+   making it hard to provide useful tracebacks and feedback if one of them
+   errors.
+
+2. pluggy.callers._multicall does not wrap the actual call to hook
+   implementations (``res = hook_impl.function(*args)``) in a try/except,
+   meaning that if *any* of the implementations throw an error, the whole hook
+   chain fails.
+
+Example
+-------
+Calling hooks the normal way:
+>>> result = plugin_manager.hook.my_hook(arg=arg)
+
+is identical to calling them with execute_hook as follows:
+>>> result = execute_hook(plugin_manager.hook.my_hook, arg=arg)
+
+If you want to know which implementation is responsible for each result, use:
+
+>>> # NOTE: this assumes plugin_manager.hook.my_hook has firstresult = True
+>>> # otherwise a list would be returned.
+>>> (result, imp) = execute_hook(
+...                     plugin_manager.hook.my_hook,
+...                     arg=arg,
+...                     return_impl=True,
+...                 )
+
+NOTE: A version code here was submitted to pluggy as PR #253:
+https://github.com/pytest-dev/pluggy/pull/253
+We may want to consider removing this code if that merges, but we may also just
+wish to maintain our internal hook looping logic.
+"""
+import sys
+from typing import (
+    Optional,
+    Sequence,
+    Any,
+    overload,
+    Literal,
+    List,
+    Tuple,
+    Union,
+)
+
+from pluggy.callers import HookCallError, _raise_wrapfail, _Result
+from pluggy.hooks import HookImpl, _HookCaller
+from .exceptions import PluginError
+
+
+@overload
+def _multicall(
+    hook_impls: Sequence[HookImpl],
+    caller_kwargs: dict,
+    return_impl: Literal[False],
+    firstresult: Literal[True],
+) -> Any:
+    ...
+
+
+@overload
+def _multicall(
+    hook_impls: Sequence[HookImpl],
+    caller_kwargs: dict,
+    return_impl: Literal[False],
+    firstresult: Literal[False],
+) -> List[Any]:
+    ...
+
+
+@overload
+def _multicall(
+    hook_impls: Sequence[HookImpl],
+    caller_kwargs: dict,
+    return_impl: Literal[True],
+    firstresult: Literal[True],
+) -> Tuple[Any, HookImpl]:
+    ...
+
+
+@overload
+def _multicall(
+    hook_impls: Sequence[HookImpl],
+    caller_kwargs: dict,
+    return_impl: Literal[True],
+    firstresult: Literal[False],
+) -> List[Tuple[Any, HookImpl]]:
+    ...
```

Comment:
I'm super-confused about these signatures, what's happening here?

Reply:
Totally optional, those are typing.overload hints.  Since the return type of _multicall changes depending on the arguments, I though it might be useful to be explicit about what output should be expected depending on the input.  Happy to remove!  Wouldn’t change any functionality

Reply:
Oooooh, I see now that the input *values* are changing. That's super crazy, I didn't know typing could do that. But I guess `Literal[True]` would not actually work if a variable that evaluates as True is passed? Personally I think it is too confusing to make it worthwhile. But I don't feel super strongly about it, but if you want to keep it I would add a comment to the top of the section explaining that the type of the return is changing in response to the value of the inputs, and `@overload` is taking care of that.

Reply:
yeah, `@overload` is very cool.   I agree though, it's too confusing here and I'll take it out.
but fyi, the main reason I'm generally tempted to do that sort of thing is for linting in my IDE since I've started using `mypy`... it would, for instance, show me if I'm trying to unpack a tuple that doesn't exist (cause i forgot to use `return_impl=True` or something)

Reply:
That's very clever. =) At skimage though, we're starting to see changing the number of return values in response to a kwarg value as an anti-pattern, precisely for this reason. Our proposal for this in the future is:

- `function.extra()` is a function with extra return values.
- You can only return the "normal" return, or everything, but,
- "everything" will come bundled in a NamedTuple or Dataclass, and different attributes can be None if desired.

So you can have `func(Arr) -> Arr` and `func.extra(Arr) -> Arr, Extras`.

Reply:
definitely agree with overloaded return types being an antipattern!  (even more so now since I'm type-"woke" and want to be able to trace types carefully).  I _love_ that approach you have there.